lib/ocpp/v201/charge_point.cpp

```diff
@@ -1094,36 +1165,75 @@ void ChargePoint::handle_reset_req(Call<ResetRequest> call) {
     // TODO(piet): B11.FR.09
     // TODO(piet): B11.FR.10
 
-    // TODO(piet): B12.FR.01
-    // TODO(piet): B12.FR.02
-    // TODO(piet): B12.FR.03
-    // TODO(piet): B12.FR.04
     // TODO(piet): B12.FR.05
     // TODO(piet): B12.FR.06
-    // TODO(piet): B12.FR.07
-    // TODO(piet): B12.FR.08
-    // TODO(piet): B12.FR.09
     EVLOG_debug << "Received ResetRequest: " << call.msg << "\nwith messageId: " << call.uniqueId;
     const auto msg = call.msg;
 
     ResetResponse response;
 
-    if (this->callbacks.is_reset_allowed_callback(msg.type)) {
+    // Check if there is an active transaction (on the given evse or if not
+    // given, on one of the evse's)
+    bool transaction_active = false;
+    if (msg.evseId.has_value() &&
+        this->evses.at(msg.evseId.value())->has_active_transaction()) {
+        transaction_active = true;
+    } else {
+        for (const auto& [evse_id, evse] : this->evses) {
+            if (evse->has_active_transaction()) {
+                transaction_active = true;
+                break;
+            }
+        }
+    }
+
+    if (this->callbacks.is_reset_allowed_callback(msg.evseId, msg.type)) {
         // reset is allowed
         response.status = ResetStatusEnum::Accepted;
     } else {
         response.status = ResetStatusEnum::Rejected;
     }
 
-    if (/* transaction is active */ false) {
+    if (response.status == ResetStatusEnum::Accepted && transaction_active) {
         if (msg.type == ResetEnum::OnIdle and !msg.evseId.has_value()) {
             // B12.FR.03
-        } else if (msg.type == ResetEnum::Immediate and !msg.evseId.has_value()) {
+            for (const auto &[evse_id, evse] : this->evses) {
+                if (!evse->has_active_transaction()) {
+                    // Set all connectors that do not have an active transaction
+                    // to 'Unavailable'.
+                    this->set_evse_connectors_unavailable(evse);
+                }
+            }
+
+            // B12.FR.01: We have to wait until transactions have ended.
+            this->reset_scheduled = true;
+            response.status = ResetStatusEnum::Scheduled;
+        } else if (msg.type == ResetEnum::Immediate and
+                   !msg.evseId.has_value()) {
             // B12.FR.04
+            for (const auto &[evse_id, evse] : this->evses) {
+                if (evse->has_active_transaction()) {
+                    callbacks.stop_transaction_callback(
```

Comment:
Changed it, but not with the 5s because that's up to the charging station how long that can take. Can you take a look if this is better?

Also worked on the other comment about the message queue that is sending things before the accepted bootnotification. Is it better with this change?